src/gleam/regex.gleam

```diff
@@ -0,0 +1,122 @@
+//// The matching algorithms of the library are based on the PCRE library, but not
+//// all of the PCRE library is interfaced and some parts of the library go beyond
+//// what PCRE offers. Currently PCRE version 8.40 (release date 2017-01-11) is used.
+
+import gleam/list
+import gleam/option.{Option}
+
+pub external type Regex
+
+/// The details about a particular match:
+///
+/// - match — the full string of the match.
+/// - index — the index of the match in the original string.
+/// - number — each match is numbered, starting from 1
+/// - submatches — a Regex can have subpatterns, sup-parts that are in parentheses.
+///
+pub type Match {
+  Match(
+    match: String,
+    index: Int,
+    number: Int,
+    submatches: List(Option(String)),
+  )
+  Error(Nil)
+}
+
+/// Create a new Regex.
+///
+/// ## Examples
+///
+///    > let re = from_string("[0-9]")
+///    >   |> result.unwrap(never())
+///    > match(re, "abc123")
+///    True
+///
+///    > match(re, "abcxyz")
+///    False
+///
+pub external fn from_string(String) -> Result(Regex, Nil) =
+  "gleam_stdlib" "regex_from_string"
+
+/// A regular expression that never matches any string.
+pub external fn never() -> Regex =
+  "gleam_stdlib" "regex_never"
+
+/// Returns a boolean indicating whether there was a match or not.
+///
+/// ## Examples
+///
+///    > let re = from_string("^f.o.?")
+///    >   |> result.unwrap(never())
+///    > match(re, "foo")
+///    True
+///
+///    > match(re, "boo")
+///    False
+///
+pub external fn match(Regex, String) -> Bool =
+  "gleam_stdlib" "regex_match"
+
+/// Split a string
+///
+/// ## Examples
+///
+///    > let re = from_string(" *, *")
+///    >   |> result.unwrap(never())
+///    > match(re, "foo,32, 4, 9  ,0")
+///    ["foo", "32", "4", "9", "0"]
+///
+pub external fn split(Regex, String) -> List(String) =
+  "gleam_stdlib" "regex_split"
+
+type MatchTuple =
+  tuple(String, Int, Int, List(Option(String)))
+
+external fn erl_find(Regex, String) -> List(MatchTuple) =
+  "gleam_stdlib" "regex_find"
+
+fn match_tuple_to_match(m: MatchTuple) -> Match {
+  case m {
+    tuple(
+      match,
+      index,
+      number,
+      submatches,
+    ) -> Match(
+      match: match,
+      index: index,
+      number: number,
+      submatches: submatches,
+    )
+    _ -> Error(Nil)
+  }
+}
+
+/// Find matches in a string
+///
+/// ## Examples
+///
+///    > let re = from_string(" *, *")
+///    > let re = regex.from_string("[oi]n a (\\w+)")
+///    >   |> result.unwrap(or: regex.never())
+///    > regex.find(re, "I am on a boat in a lake.")
+///    [
+///      Match(
+///        match: "on a boat",
+///        index: 5,
+///        number: 1,
+///        submatches: [Some("boat")]
+///      ),
+///      Match(
+///        match: "in a lake",
+///        index: 15,
+///        number: 2,
+///        submatches: [Some("lake")]
+///      )
+///    ]
+pub fn find(re: Regex, str: String) -> List(Match) {
```

Comment:
I'm not sure about this name, is it taken from somewhere?

Reply:
It's called `scan` in Elixir, Ruby and probably some other languages, `find` seems to be confusing, I changed it to `scan` instead.